### PR TITLE
KEP-1321: Send state requests when swarm checkpoint conflicts with local

### PR DESCRIPTION
**Pull request by isabelsavannah**
checkpoint. Refactor checkpoint handling.

### CODE REVIEW

**Comment by isabelsavannah** [comment]
github isn't detecting the rename, but mocks/smart_mock_io.cpp is just slightly modified from old node/test/node_test_common.cpp.

File: pbft/pbft_checkpoint_manager.cpp [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+    return result;
+}
+
+size_t
+pbft_checkpoint_manager::partial_checkpoint_proofs_count() const
+{
+    std::lock_guard<std::mutex> lock(this->lock);
+    return this->partial_checkpoint_proofs.size();
+
+}
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: compilers don't dig modules with no empty space at the back. Meh, fix it later...

File: pbft/pbft_checkpoint_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+#include <pbft/pbft_checkpoint_manager.hpp>
+#include <proto/pbft.pb.h>
+#include <utils/bytes_to_debug_string.hpp>
+#include <utils/make_endpoint.hpp>
+#include <algorithm>
+#include <pbft/pbft.hpp>
+#include <iterator>
+
+namespace
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: I've read through this module and I think it is mostly extracted from pbft. So the functionality must be good. 

The code formatting looks very nice as well. I think the re factoring is good.

File: pbft/pbft_checkpoint_manager.hpp [hunk 3]
```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2018 Bluzelle
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.hpp]: A very nice extraction.

File: pbft/pbft_checkpoint_manager.cpp [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+    if (checkpoint_attestants >= pbft::honest_member_size(peers->size()))
+    {
+        this->stabilize_checkpoint(cp);
+
+        if (this->latest_local_checkpoint.value().first < cp.first)
+        {
+            this->send_delayed_state_request(cp);
+        }
+        else if (this->latest_local_checkpoint.value().first == cp.first
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: I think this is one place where we do the actual test of KEP-1321

File: pbft/pbft_checkpoint_manager.cpp [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    bzn_envelope msg;
+    msg.set_checkpoint_msg(cp_msg.SerializeAsString());
+
+    auto msg_ptr = std::make_shared<bzn_envelope>(msg);
+    for (const auto& peer : *(this->config_store->current()->get_peers()))
+    {
+        this->node->send_signed_message(make_endpoint(peer), msg_ptr);
+    }
+
+    if (cp.first == this->latest_local_checkpoint.value().first && cp.second != this->latest_local_checkpoint.value().second)
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: KEP-1321 sending a state request when the local checkpoint is wrong. cool

File: pbft/pbft_checkpoint_manager.cpp [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    {
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: KEP-1321 cool, verifying that the local checkpoint matches the swarm, sending a state request if it does not...

File: pbft/pbft_checkpoint_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        this->stabilize_checkpoint(cp);
+
+        if (this->latest_local_checkpoint.value().first < cp.first)
+        {
+            this->send_delayed_state_request(cp);
+        }
+        else if (this->latest_local_checkpoint.value().first == cp.first
+                 && this->latest_stable_checkpoint.value().second != cp.second)
+        {
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: KEP-1321 verifying here as well, cool.

File: pbft/pbft_checkpoint_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+            [cp, weak_this = weak_from_this()](auto ec)
+            {
+                if (ec)
+                {
+                    return;
+                }
+
+                auto strong_this = weak_this.lock();
+
+                if (strong_this && strong_this->get_latest_stable_checkpoint() == cp)
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: KEP-1321 cool, and here. I think these are the instances of the new functionality for this issue.

File: pbft/pbft_checkpoint_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+                    std::lock_guard<std::mutex> lock(strong_this->lock);
+                    strong_this->send_state_request();
+                }
+            }
+    );
+}
+
+void
+pbft_checkpoint_manager::send_state_request()
+{
```
Comment (rnistuk) [inline pbft/pbft_checkpoint_manager.cpp]: KEP-1321 more of the functionality specific to this issue. Sends off a get state request to random (!) peer?

That's cool. Isabel tells me that the request is expensive. And I think that if the random peer is down or something, this procedure will just happen again sometime in the future.

File: pbft/pbft_config_store.cpp [hunk 10]
```diff
@@ -46,19 +54,22 @@ pbft_config_store::get(const hash_t& hash) const
...
 {
+    std::lock_guard<std::mutex> lock(this->lock);
     auto it = this->view_configs.find(view);
-    return it == this->view_configs.end() ? nullptr : this->get(it->second.value());
+    return it == this->view_configs.end() ? nullptr : this->private_get(it->second.value());
 }
 
 bool
 pbft_config_store::set_prepared(const hash_t& hash)
 {
```
Comment (rnistuk) [inline pbft/pbft_config_store.cpp]: Ah, adding some locks for thread safety, cool

**Pull request merged**